File: gherkin/python/gherkin/ast_builder.py
from .ast_node import AstNode
from .errors import AstBuilderException
from .stream.id_generator import IdGenerator

class AstBuilder(object):
    def __init__(self, id_generator=None):
        self.id_generator = id_generator
        if self.id_generator is None:
            self.id_generator = IdGenerator()
        self.reset()

    def reset(self):
        self.stack = [AstNode('None')]
        self.comments = []
        self.id_counter = 0

    def start_rule(self, rule_type):
        self.stack.append(AstNode(rule_type))

    def end_rule(self, rule_type):
        node = self.stack.pop()
        self.current_node.add(node.rule_type, self.transform_node(node))

    def build(self, token):
        if token.matched_type == 'Comment':
            self.comments.append({
                'location': self.get_location(token),
                'text': token.matched_text
            })
        else:
            self.current_node.add(token.matched_type, token)

    def get_result(self):
        return self.current_node.get_single('GherkinDocument')

    @property
    def current_node(self):
        return self.stack[-1]

    def get_location(self, token, column=None):
        return (token.location if not column else
                {'line': token.location['line'], 'column': column})

    def get_tags(self, node):
        tags = []
        tags_node = node.get_single('Tags')
        if not tags_node:
            return tags

        for token in tags_node.get_tokens('TagLine'):
            tags += [{'id': self.id_generator.get_next_id(),
                      'location': self.get_location(token, tag_item['column']),
                      'name': tag_item['text']} for tag_item in token.matched_items]

        return tags

    def get_table_rows(self, node):
        rows = [{'id': self.id_generator.get_next_id(),
                 'location': self.get_location(token),
                 'cells': self.get_cells(token)} for token in node.get_tokens('TableRow')]
        self.ensure_cell_count(rows)
        return rows

    def ensure_cell_count(self, rows):
        if not rows:
            return

        cell_count = len(rows[0]['cells'])
        for row in rows:
            if len(row['cells']) != cell_count:
                raise AstBuilderException("inconsistent cell count within the table",
                                          row['location'])

    def get_cells(self, table_row_token):
        return [self.reject_nones(
            {'location': self.get_location(table_row_token, cell_item['column']),
             'value': cell_item['text']}) for cell_item in table_row_token.matched_items]

    def get_description(self, node):
        return node.get_single('Description', '')

    def get_steps(self, node):
        return node.get_items('Step')

    def transform_node(self, node):
        if node.rule_type == 'Step':
            step_line = node.get_token('StepLine')
            step_argument_type = 'dummy_type'
            step_argument = None
            if node.get_single('DataTable'):
                step_argument_type = 'dataTable'
                step_argument = node.get_single('DataTable')
            elif node.get_single('DocString'):
                step_argument_type = 'docString'
                step_argument = node.get_single('DocString')

            return self.reject_nones({
                'id': self.id_generator.get_next_id(),
                'location': self.get_location(step_line),
                'keyword': step_line.matched_keyword,
                'text': step_line.matched_text,
                step_argument_type: step_argument
            })
        elif node.rule_type == 'DocString':
            separator_token = node.get_tokens('DocStringSeparator')[0]
            media_type = (separator_token.matched_text if len(separator_token.matched_text) > 0
                            else None)
            line_tokens = node.get_tokens('Other')
            content = '\n'.join([t.matched_text for t in line_tokens])

            return self.reject_nones({
                'location': self.get_location(separator_token),
                'content': content,
                'delimiter': separator_token.matched_keyword,
                'mediaType': media_type
            })
        elif node.rule_type == 'DataTable':
            rows = self.get_table_rows(node)
            return self.reject_nones({
                'location': rows[0]['location'],
                'rows': rows,
            })
        elif node.rule_type == 'Background':
            background_line = node.get_token('BackgroundLine')
            description = self.get_description(node)
            steps = self.get_steps(node)

            return self.reject_nones({
                'id': self.id_generator.get_next_id(),
                'location': self.get_location(background_line),
                'keyword': background_line.matched_keyword,
                'name': background_line.matched_text,
                'description': description,
                'steps': steps
            })
        elif node.rule_type == 'ScenarioDefinition':
            tags = self.get_tags(node)
            scenario_node = node.get_single('Scenario')
            scenario_line = scenario_node.get_token('ScenarioLine')
            description = self.get_description(scenario_node)
            steps = self.get_steps(scenario_node)
            examples = scenario_node.get_items('ExamplesDefinition')

            return self.reject_nones({
                'id': self.id_generator.get_next_id(),
                'tags': tags,
                'location': self.get_location(scenario_line),
                'keyword': scenario_line.matched_keyword,
                'name': scenario_line.matched_text,
                'description': description,
                'steps': steps,
                'examples': examples
            })
        elif node.rule_type == 'ExamplesDefinition':
            tags = self.get_tags(node)
            examples_node = node.get_single('Examples')
            examples_line = examples_node.get_token('ExamplesLine')
            description = self.get_description(examples_node)
            examples_table_rows = examples_node.get_single('ExamplesTable')
            table_header = examples_table_rows[0] if examples_table_rows else None
            table_body = examples_table_rows[1:] if examples_table_rows else []

            return self.reject_nones({
                'id': self.id_generator.get_next_id(),
                'tags': tags,
                'location': self.get_location(examples_line),
                'keyword': examples_line.matched_keyword,
                'name': examples_line.matched_text,
                'description': description,
                'tableHeader': table_header,
                'tableBody': table_body
            })
        elif node.rule_type == 'ExamplesTable':
            return self.get_table_rows(node)
        elif node.rule_type == 'Description':
            line_tokens = node.get_tokens('Other')
            # Trim trailing empty lines
            last_non_empty = next(i for i, j in reversed(list(enumerate(line_tokens)))
                                  if j.matched_text)
            description = '\n'.join([token.matched_text for token in
                                     line_tokens[:last_non_empty + 1]])

            return description
        elif node.rule_type == 'Rule':
            header = node.get_single('RuleHeader')
            if not header:
                return

            tags = self.get_tags(header)
            rule_line = header.get_token('RuleLine')
            if not rule_line:
                return

            children = []
            background = node.get_single('Background')
            if (background):
                children.append({'background': background})
            children = children + [{'scenario': i} for i in node.get_items('ScenarioDefinition')]
            description = self.get_description(header)

            return self.reject_nones({
                'id': self.id_generator.get_next_id(),
                'tags': tags,
                'location': self.get_location(rule_line),
                'keyword': rule_line.matched_keyword,
                'name': rule_line.matched_text,
                'description': description,
                'children': children
            })
        elif node.rule_type == 'Feature':
            header = node.get_single('FeatureHeader')
            if not header:
                return

            tags = self.get_tags(header)
            feature_line = header.get_token('FeatureLine')
            if not feature_line:
                return

            children = []
            background = node.get_single('Background')
            if (background):
                children.append({'background': background})
            children = children + [{'scenario': i} for i in node.get_items('ScenarioDefinition')]
            children = children + [{'rule': i} for i in node.get_items('Rule')]
            description = self.get_description(header)
            language = feature_line.matched_gherkin_dialect

            return self.reject_nones({
                'tags': tags,
                'location': self.get_location(feature_line),
                'language': language,
                'keyword': feature_line.matched_keyword,
                'name': feature_line.matched_text,
                'description': description,
                'children': children
            })
        elif node.rule_type == 'GherkinDocument':
            feature = node.get_single('Feature')

            return self.reject_nones({
                'feature': feature,
                'comments': self.comments
            })
        else:
            return node

    def reject_nones(self, values):
        return {k: v for k, v in values.items() if v is not None}

File: gherkin/ruby/lib/gherkin/ast_builder.rb
require 'cucumber/messages'
require_relative 'ast_node'

module Gherkin
  class AstBuilder
    def initialize(id_generator)
      @id_generator = id_generator
      reset
    end

    def reset
      @stack = [AstNode.new(:None)]
      @comments = []
    end

    def start_rule(rule_type)
      @stack.push AstNode.new(rule_type)
    end

    def end_rule(rule_type)
      node = @stack.pop
      current_node.add(node.rule_type, transform_node(node))
    end

    def build(token)
      if token.matched_type == :Comment
        @comments.push(Cucumber::Messages::Comment.new(
          location: get_location(token, 0),
          text: token.matched_text
        ))
      else
        current_node.add(token.matched_type, token)
      end
    end

    def get_result
      current_node.get_single(:GherkinDocument)
    end

    def current_node
      @stack.last
    end

    def get_location(token, column)
      column = column == 0 ? token.location[:column] : column
      Cucumber::Messages::Location.new(
        line: token.location[:line],
        column: column
      )
    end

    def get_tags(node)
      tags = []
      tags_node = node.get_single(:Tags)
      return tags unless tags_node

      tags_node.get_tokens(:TagLine).each do |token|
        token.matched_items.each do |tag_item|
          tags.push(Cucumber::Messages::Tag.new(
            location: get_location(token, tag_item.column),
            name: tag_item.text,
            id: @id_generator.new_id
          ))
        end
      end

      tags
    end

    def get_table_rows(node)
      rows = node.get_tokens(:TableRow).map do |token|
        Cucumber::Messages::TableRow.new(
          id: @id_generator.new_id,
          location: get_location(token, 0),
          cells: get_cells(token)
        )
      end
      ensure_cell_count(rows)
      rows
    end

    def ensure_cell_count(rows)
      return if rows.empty?
      cell_count = rows[0].cells.length
      rows.each do |row|
        if row.cells.length != cell_count
          raise AstBuilderException.new("inconsistent cell count within the table", row.location.to_h)
        end
      end
    end

    def get_cells(table_row_token)
      table_row_token.matched_items.map do |cell_item|
        Cucumber::Messages::TableCell.new(
          location: get_location(table_row_token, cell_item.column),
          value: cell_item.text
        )
      end
    end

    def get_description(node)
      node.get_single(:Description) || ''
    end

    def get_steps(node)
      node.get_items(:Step)
    end

    def transform_node(node)
      case node.rule_type
      when :Step
        step_line = node.get_token(:StepLine)
        data_table = node.get_single(:DataTable)
        doc_string = node.get_single(:DocString)

        step = Cucumber::Messages::Step.new(
          location: get_location(step_line, 0),
          keyword: step_line.matched_keyword,
          text: step_line.matched_text,
          data_table: data_table,
          doc_string: doc_string,
          id: @id_generator.new_id
        )
      when :DocString
        separator_token = node.get_tokens(:DocStringSeparator)[0]
        media_type = separator_token.matched_text == '' ? nil : separator_token.matched_text
        line_tokens = node.get_tokens(:Other)
        content = line_tokens.map { |t| t.matched_text }.join("\n")

        Cucumber::Messages::DocString.new(
          location: get_location(separator_token, 0),
          content: content,
          delimiter: separator_token.matched_keyword,
          media_type: media_type
        )
      when :DataTable
        rows = get_table_rows(node)
        Cucumber::Messages::DataTable.new(
          location: rows[0].location,
          rows: rows
        )
      when :Background
        background_line = node.get_token(:BackgroundLine)
        description = get_description(node)
        steps = get_steps(node)

        Cucumber::Messages::Background.new(
          id: @id_generator.new_id,
          location: get_location(background_line, 0),
          keyword: background_line.matched_keyword,
          name: background_line.matched_text,
          description: description,
          steps: steps
        )
      when :ScenarioDefinition
        tags = get_tags(node)
        scenario_node = node.get_single(:Scenario)
        scenario_line = scenario_node.get_token(:ScenarioLine)
        description = get_description(scenario_node)
        steps = get_steps(scenario_node)
        examples = scenario_node.get_items(:ExamplesDefinition)
        Cucumber::Messages::Scenario.new(
          id: @id_generator.new_id,
          tags: tags,
          location: get_location(scenario_line, 0),
          keyword: scenario_line.matched_keyword,
          name: scenario_line.matched_text,
          description: description,
          steps: steps,
          examples: examples
        )
      when :ExamplesDefinition
        tags = get_tags(node)
        examples_node = node.get_single(:Examples)
        examples_line = examples_node.get_token(:ExamplesLine)
        description = get_description(examples_node)
        rows = examples_node.get_single(:ExamplesTable)

        table_header = rows.nil? ? nil : rows.first
        table_body = rows.nil? ? [] : rows[1..-1]

        Cucumber::Messages::Examples.new(
          id: @id_generator.new_id,
          tags: tags,
          location: get_location(examples_line, 0),
          keyword: examples_line.matched_keyword,
          name: examples_line.matched_text,
          description: description,
          table_header: table_header,
          table_body: table_body
        )
      when :ExamplesTable
        get_table_rows(node)
      when :Description
        line_tokens = node.get_tokens(:Other)
        # Trim trailing empty lines
        last_non_empty = line_tokens.rindex { |token| !token.line.trimmed_line_text.empty? }
        description = line_tokens[0..last_non_empty].map { |token| token.matched_text }.join("\n")
        return description
      when :Feature
        header = node.get_single(:FeatureHeader)
        return unless header
        tags = get_tags(header)
        feature_line = header.get_token(:FeatureLine)
        return unless feature_line
        children = []
        background = node.get_single(:Background)
        children.push(Cucumber::Messages::FeatureChild.new(background: background)) if background
        node.get_items(:ScenarioDefinition).each do |scenario|
          children.push(Cucumber::Messages::FeatureChild.new(scenario: scenario))
        end
        node.get_items(:Rule).each do |rule|
          children.push(Cucumber::Messages::FeatureChild.new(rule: rule))
        end
        description = get_description(header)
        language = feature_line.matched_gherkin_dialect

        Cucumber::Messages::Feature.new(
          tags: tags,
          location: get_location(feature_line, 0),
          language: language,
          keyword: feature_line.matched_keyword,
          name: feature_line.matched_text,
          description: description,
          children: children,
        )
      when :Rule
        header = node.get_single(:RuleHeader)
        return unless header
        rule_line = header.get_token(:RuleLine)
        return unless rule_line
        tags = get_tags(header)
        children = []
        background = node.get_single(:Background)
        children.push(Cucumber::Messages::RuleChild.new(background: background)) if background
        node.get_items(:ScenarioDefinition).each do |scenario|
          children.push(Cucumber::Messages::RuleChild.new(scenario: scenario))
        end
        description = get_description(header)

        Cucumber::Messages::Rule.new(
          id: @id_generator.new_id,
          tags: tags,
          location: get_location(rule_line, 0),
          keyword: rule_line.matched_keyword,
          name: rule_line.matched_text,
          description: description,
          children: children,
        )
      when :GherkinDocument
        feature = node.get_single(:Feature)
        Cucumber::Messages::GherkinDocument.new(
          comments: @comments,
          feature: feature
        )
      else
        return node
      end
    end
  end
end
